### PR TITLE
[libcgal-julia] Update to v0.16.0 w/ libjulia v1.3.1

### PR DESCRIPTION
**Pull request by rgcv**
Supersedes @fingolfin's draft #2326, betting on supporting julia versions increasingly since the latest LTS.

### CODE REVIEW

**Comment by rgcv** [comment]
Out of curiosity, I'm seeing in the build logs that the following dependency versions are being pulled in:

- libcxxwrap_julia v0.8.5
- GMP v6.2.0
- MPFR v4.1.1

whereas I would be expecting versions v0.8.3, v6.1.2, and v4.0.2 respectively.

Now I haven't been around a while and I may or may not have consequently have forgotton how BinaryBuilder and julia's packaging system works ^^ but I'm guessing that since these versions are being pulled in, I should somehow strictly specify the versions I need?  Despite undesirable, this would be fine for GMP and MPFR, specifying them as build dependencies.  However, it wouldn't go well for libcxxwrap_julia (sort of, one could rely on the julia minor version convention and pull that in from the separate build scripts).

Am I getting concerned over nothing?  Or is this in fact an issue that should be addressed?

**Comment by fingolfin** [comment]
Good catch -- adding versioned dependencies for MPFR and GMP may be necessary... it depends on whether or not libcgal_julia also directly links against either of them... this could be determined by inspecting the result of the macOS build. I can try to do this, but no promises as I also have to take care of the kids right now and may not have a chance ... ;-)

If it *does* link against either, the binaries produced here won't be usable at all in Julia 1.3-1.5. Unfortunately the special build version restriction is not "inherited".... I guess we should go through all JLL recipes that depend on GMP and/or MPFR now, then check which JLL depend on them, and decided whether they need to get explicit (build)dependencies on either GMP or MPFR, and then repeat this recursively for JLLs depending on those, etc.. All in all this doesn't scale too well. Hrm

There is no need for similar tricks with libcxxwrap_julia, though, as there the underlying dylibs really have identical versions, and are fully ABI compatible.

**Comment by fingolfin** [comment]
OK, just had a chance to check. Yes, you do need to add
```julia
    BuildDependency(PackageSpec(name="GMP_jll", version=v"6.1.2")),
    BuildDependency(PackageSpec(name="MPFR_jll", version=v"4.0.2")),
```

File: L/libcgal_julia/common.jl [hunk 1]
```diff
@@ -60,24 +53,25 @@ install_license $jlcgaldir/LICENSE
...
 
 # The products that we will ensure are always built
 products = [
     LibraryProduct("libcgal_julia_exact", :libcgal_julia_exact),
     LibraryProduct("libcgal_julia_inexact", :libcgal_julia_inexact),
 ]
 
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    BuildDependency(PackageSpec(name="libjulia_jll", version=julia_version)),
```
Comment (fingolfin) [inline L/libcgal_julia/common.jl:68]:
```suggestion
    BuildDependency(PackageSpec(name="libjulia_jll", version=julia_version)),
    BuildDependency(PackageSpec(name="GMP_jll", version=v"6.1.2")),
    BuildDependency(PackageSpec(name="MPFR_jll", version=v"4.0.2")),

```

File: L/libcgal_julia/libcgal_julia@1.3/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,2 @@
+julia_version = v"1.3.1"
+include(joinpath(dirname(@__DIR__), "common.jl"))
```
[the remark this answers is not in the document]
Comment (rgcv) [reply L/libcgal_julia/libcgal_julia@1.3/build_tarballs.jl:2]: Not really, just a habit I grew over time ^^

**Comment by rgcv** [comment]
Apologies for stressing the pipeline, force pushed reworded commit messages to preserve consistency.

**Comment by rgcv** [comment]
Builds all go through!  A few warnings arise about minimum instruction set in mingw32 builds being avx2 vs. x86_64, as well as some symlinks of certain objects present in different artifacts for linux-musl (libz.so) and darwin (set_soname_libgcc_ext.dylib,set_soname_libgcc_s.dylib) builds.  Other than that, nothing major.

**Review by fingolfin — APPROVED** [review]
This looks good from my POV.

File: L/libcgal_julia/common.jl [hunk 3]
```diff
@@ -60,24 +53,27 @@ install_license $jlcgaldir/LICENSE
...
+    BuildDependency(PackageSpec(name="GMP_jll", version=v"6.1.2")),
+    BuildDependency(PackageSpec(name="MPFR_jll", version=v"4.0.2")),
+
```
Comment (giordano) [inline L/libcgal_julia/common.jl:71]: I don't understand why this is necessary: if GMP/MPFR aren't direct dependencies of the library product, why do the versions matter?  I probably miss the results of @fingolfin investigation which lead to https://github.com/JuliaPackaging/Yggdrasil/pull/2333#issuecomment-752942350

Comment (fingolfin) [reply L/libcgal_julia/common.jl:71]: They *are* direct dependencies of the library product, as it turns out (that was the outcome of my investigation). We just didn't have to declare them explicitly so far, as they were pulled in by CGAL_jll. But that pull int the latest versions it can install, so 6.2.0 resp 4.1.0, which is bad.

Comment (giordano) [reply L/libcgal_julia/common.jl:71]: So they are not `BuildDependency` and can do
```suggestion
    Dependency("GMP_jll", v"6.1.2"),
    Dependency("MPFR_jll", v"4.0.2"),

```
?  Or am I missing something else?

Comment (rgcv) [reply L/libcgal_julia/common.jl:71]: What I don't seem to understand is why pulling `CGAL_jll` v5.2, which depends on GMP v6.1.2 and MPFR v4.0.2, won't pull these versions of GMP/MPFR transitively, pulling in the latest versions instead.  I never really specified versions or dependencies for GMP/MPFR in `libcgal-julia` since what the wrapper depends on is `CGAL`, *implicitly* depending on GMP/MPFR.  Specifying these explicitly seems odd and wrong, hence why I thought, as per @fingolfin's suggestion, a `BuildDependency` would make more sense in this case since there isn't an explicit hard-dependency on GMP nor MPFR.

Comment (rgcv) [reply L/libcgal_julia/common.jl:71]: Just tried adding `CGAL_jll` on its own locally.  While it pulled `GMP_jll` v6.1.2+6, it pulled `MPFR_jll` v4.1.0+1.  Tested on a Windows 10 x64 machine running julia 1.5.

Comment (giordano) [reply L/libcgal_julia/common.jl:71]: > What I don't seem to understand is why pulling `CGAL_jll` v5.2, which depends on GMP v6.1.2 and MPFR v4.0.2, won't pull these resp. versions transitively, pulling in the latest versions instead.

Uh?  There is no such restriction in the registry: https://github.com/JuliaRegistries/General/blob/49c116c08da545b877943783652a6f340c0ff88e/C/CGAL_jll/Compat.toml.  Then it's normal that the latest versions of GMP and MPFR are used.

Comment (fingolfin) [reply L/libcgal_julia/common.jl:71]: @rgcv I think the confusion is from the meaning of the syntax `Dependency("GMP_jll", v"6.1.2"),` -- this does in fact *not* impose a compat restriction of the dependency "GMP_jll" of the JLL package being produced in the end -- rather, that version is only used by BinaryBuilder to choose which version of GMP_jll to install *while building the JLL*; it does not affect which version of the dependency is later installed at "run time".

Comment (fingolfin) [reply L/libcgal_julia/common.jl:71]: As to whether this should be declared as a BuildDependency or Dependency in the two given ways: Both are de-facto equivalent in practice, it doesn't really matter. The only difference is whether GMP_jll/MPFR_jll will be listed as direct dependencies; but since they are indirect dependencies, it doesn't matter.

I should also point out that all I checked is whether the shared libraries in libcgal_julia_jll explicitly link against libgmp / libmpfr (i.e., by invoking `otool -L` on them on macOS). And it turned out that this *is* the case. However, I did not dig further *why* it is; there are at least two possibilities (separately for each of GMP and MPFR):
1. They are "true" dependencies as in: the shared libraries in `libcgal_julia_jll` really invoke GMP resp. MPFR APIs (directly and explicitly in its C++ code, or indirectly, by invoking CGAL macros or static inline functions (declared in headers) that do so; or
2. This is a false dependency, only caused by `-lgmp -lmpfr` linker flags that get injected somewhere, but which are not actually needed.

In the latter case, one could also resolve this by adjusting the build system to avoid these linker flags, then drop GMP_jll resp. MPFR_jll as a direct (build) dependency. Alas, I won't try to do that, it's way out of scope for me (i.e., I don't use CGAL or libcgal_julia or anything related to them ;-) ), but I wanted to mention it to fully explain the situation as I see it.

Comment (rgcv) [reply L/libcgal_julia/common.jl:71]: > There is no such restriction in the registry: JuliaRegistries/General@49c116c/C/CGAL_jll/Compat.toml. Then it's normal that the latest versions of GMP and MPFR are used.

> I think the confusion is from the meaning of the syntax Dependency("GMP_jll", v"6.1.2"), -- this does in fact not impose a compat restriction of the dependency "GMP_jll" of the JLL package being produced in the end

Yes, Max hit the nail in the head.  I thought the specified version would impose a strict compatibility restriction on the version of the artifact the jll pulls in.  My bad.

Okay, so.  Hmm.  Despite me building CGAL in non-header only mode to generate the shared libraries, when building libcgal_julia, it is built against CGAL in header-only mode, essentially compiling in CGAL code that ends up referencing GMP/MPFR code.  This might explain why libcgal_julia links against GMP/MPFR.  I.e., something along these lines

> They are "true" dependencies as in: the shared libraries in libcgal_julia_jll really invoke GMP resp. MPFR APIs (directly and explicitly in its C++ code, or indirectly, by invoking CGAL macros or static inline functions (declared in headers) that do so

So now I'm not sure if the correct approach would be specifying the `BuildDependency`s for GMP/MPFR while building libcgal_julia or specifying the GMP/MPFR versions CGAL is compatible with.  That would require a rebuild of `CGAL_jll`, which is annoying, but it removes the nuisance of having to "break abstraction" *per se* and re-specifying GMP and MPFR dependency versions despite libcgal_julia not referencing either directly.

Comment (rgcv) [reply L/libcgal_julia/common.jl:71]: I was thinking about submitting a PR for rebuilding CGAL with stricter compat on GMP and MPFR.  However, one could argue that the strict versions for GMP and MPFR are due to the presence of libjulia, a build dependency required by libcgal_julia, not CGAL itself (which works with any of the latest versions available, in theory).  I'm still not sure if this is reason enough to specify the hard dependencies here as this still relies on the explicit knowledge that CGAL depends on these same dependencies.  It exhibits too tight a coupling for my taste, but might just me being anal...  Apologies for revealing so many nit-picky concerns.

Comment (rgcv) [reply L/libcgal_julia/common.jl:71]: @fingolfin Sorry to bug you about this, but I have a question

> that version is only used by BinaryBuilder to choose which version of GMP_jll to install while building the JLL; it does not affect which version of the dependency is later installed at "run time"

So there wouldn't be a need for specifying GMP/MPFR's versions when building `CGAL_jll`, right?  That restriction should only then be applied when building `libcgal_julia_jll`, correct?  This could lead to "loosening" the build versions of the dependencies when building `CGAL_jll` and maybe (slightly) warrant the need to add `BuildDependency`s for `libcgal_julia`.

Apologies, once again, for any and all inconvenience that I may be causing.

Comment (giordano) [reply L/libcgal_julia/common.jl:71]: I must say I'm a bit lost.  I think there are two options:

1. `libcgal_julia` directly links to GMP/MPFR at runtime: this recipe should have `Dependency("GMP_jll", v"...")`... (not `BuildDependency`)
2. `libcgal_julia` doesn't directly link to GMP/MPFR at runtime: no need to specify any of this dependency.

My understanding is that we are in 1.  I don't understand the need to change anything in `CGAL`

Comment (fingolfin) [reply L/libcgal_julia/common.jl:71]: No!

The runtime version of GMP_jll and MPFR_jll are determined by the Julia version. You must not restrict them in your JLL, unless you really want your JLL be limit to a subset of all Julia versions.

On the other hand, during build time, we must use the oldest version of each -- if a newer one is used during build time, the at least on macOS the resulting binaries won't be usable with any older versions of libgmp / libmpfr.

tl;dr: everything is alright now, don't mess with it

Comment (rgcv) [reply L/libcgal_julia/common.jl:71]: @giordano Indeed `libcgal_julia` links against GMP and MPFR by way of compiling against CGAL, a header-only lib that potentially inlines GMP/MPFR related code that is then linked against.  However, `libcgal_julia` doesn't *directly* depend on GMP/MPFR, which makes it odd to have it depend on them in any way, be them build-time or runtime dependencies.  Those should, in my opinion, be pulled by CGAL_jll.

In order to get the correct ones, I'd have to specify compat versions, not just build versions (i.e. use `PackageSpec` to specify the version vs. `Dependency`'s regular constructor).  However, that would *tighten* the versions of GMP and MPFR CGAL_jll is compatible with when those specific versions are required by libjulia, something libcgal_julia compiles against, not CGAL.

So my confusion is mostly an inner conflict of where should these dependencies be strictly declared.  On one side, it makes sense to have CGAL_jll pull the right versions because libcgal_julia doesn't *explicitly* use either.  On the other hand, those restrictions are imposed by libjulia which CGAL shouldn't know about, hence these restrictions could be applied to libcgal_julia_jll, not CGAL_jll.  However, the latter makes indirect implicit dependencies explicit, in a sense, which I dislike since, say, for example, CGAL stops depending on GMP or MPFR.  Then libcgal_julia would have to remove those "hard" dependencies as well.

It is true that libcgal_julia and CGAL are very tightly coupled, so it isn't much of an issue specifying the versions of GMP and MPFR that libjulia *likes* in the buildscript for `libcgal_julia_jll`, i.e. here.

I'm not sure that clears up what I'm trying to say.  I think I'm sort of repeating myself at this point.  Nonetheless, I don't think it makes much sense for `libcgal_julia_jll` depend on GMP and MPFR during runtime.  Then again, by specifying them as `BuildDependency`s, that wouldn't pull the correct GMP/MPFR jlls, I would imagine.

Comment (rgcv) [reply L/libcgal_julia/common.jl:71]: > No!

Assuming that was at me! ^^

So, `CGAL_jll` and `libcgal_julia_jll` stay as is.  Still a bit offputting to respecify GMP and MPFR, but I'll take it given the aforementioned very tight coupling between CGAL and libcgal_julia.
